### ISSUE TITLE
Support for the 'inner' method in the Vector class

### ISSUE THREAD

**Issue by ctjacobs** [firedrakeproject/firedrake#475]
In DOLFIN the following code

```
from dolfin import *
mesh = UnitSquareMesh(1, 1)
vfs = VectorFunctionSpace(mesh, "CG", 2)
fs = FunctionSpace(mesh, "CG", 1)
mfs = MixedFunctionSpace([vfs, fs])

f = project(Expression(("1.0", "2.0", "3.0")), mfs)
g = project(Expression(("1.0", "2.0", "3.0")), mfs)

print f.vector().array()
print g.vector().array()

print f.vector().inner(g.vector())
```

gives

```
Solving linear system of size 22 x 22 (PETSc Krylov solver).
Solving linear system of size 22 x 22 (PETSc Krylov solver).
[ 1.  1.  1.  1.  2.  2.  2.  3.  3.  1.  1.  2.  2.  2.  3.  1.  1.  1.
  2.  2.  2.  3.]
[ 1.  1.  1.  1.  2.  2.  2.  3.  3.  1.  1.  2.  2.  2.  3.  1.  1.  1.
  2.  2.  2.  3.]
81.0
```

In Firedrake this fails because the `inner` method does not exist in the `Vector` class:

```
pyop2:INFO Solving linear variational problem...
pyop2:INFO Solving linear variational problem...done
pyop2:INFO Solving linear variational problem...
pyop2:INFO Solving linear variational problem...done
[ array([[ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.]])
 array([ 3.,  3.,  3.,  3.])]
[ array([[ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.],
       [ 1.,  2.]])
 array([ 3.,  3.,  3.,  3.])]
Traceback (most recent call last):
  File "vector_inner.py", line 13, in <module>
    print f.vector().inner(g.vector())
AttributeError: 'Vector' object has no attribute 'inner'
```

**Comment by miklos1** [comment]
sum(e.reshape(-1).dot(f.reshape(-1)) for e, f in zip(f.vector().array(), g.vector().array())) may do the job.

**Comment by wence-** [comment]
So we can support this stuff more generally by working with the PETSc vector view of the dat.  But I think we want to sit down with Christian (and maybe remotely Simon) and figure out what the right interface for this is.

**Comment by miklos1** [comment]
sum((x*y).sum() for x, y in zip(f.vector().array(), g.vector().array())) slightly shorter.

**Comment by wence-** [comment]
Incorrect in parallel.

```python
def inner(self, other):
    return self.dat.inner(other.dat)
```

**Comment by ctjacobs** [comment]
Addressed in #489. Closing.
